### PR TITLE
Fix various deadlocks on ProjectRecovery shutdown

### LINKED ISSUE #22714: Project Recovery - Race condition on ADS
Project recovery should gracefully handle workspaces being renamed in the ADS. Whilst a more permanent fix needs to be designed, in the mean time let's mitigate the effects for this release by catching the specific exception

### Steps to reproduce the behavior
- In the Mantid.Properties change saving to every second
- Run the following script
``` python
import time

ws_list = []
for i in range(30):
    ws_list.append(CreateSampleWorkspace(OutputWorkspace="ws {0}".format(i)))

while True:
    for i in range(0, 30):
        ws_list[i] = RenameWorkspace(ws_list[i], OutputWorkspace="ws-1 {0}".format(i))
        ws_list[i] = RenameWorkspace(ws_list[i], OutputWorkspace="ws-2 {0}".format(i))
    time.sleep(1)
```

An exception will be thrown in the project recovery thread, causing it to stop

### LINKED ISSUE #22736: Attempting to shutdown Mantid whilst project recovery is running triggers deadlock
**Original reporter:** @mantid-roman 

### Expected behavior
Closing Mantid whilst project recovery is saving should work gracefully

### Actual behavior
A deadlock happens on the thread mutex with the lock guard

### Steps to reproduce the behavior
Open a large workspace group with a large amount of history
Attempt to close Mantid

### Platforms affected
All

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
Fixes various deadlocks in ProjectRecovery:

- Scopes a critical section correctly, so we don't call into the GUI thread which is also waiting for the critical section
- Changes thread joining to detaching so we don't block the GUI (this also means we can get a `CancelException`
- Fixes a problem with workspaces being removed from the ADS before we are ready

**To test:**
- Open a workspace group with a large amount of history (or duplicate a workspace many times), I have an example if required
- Set the logger to debug
- Wait for project saving to start
- Before it finishes close Mantid normally
- Ensure it does not deadlock

----------------------
- Run the script attached to this issue #22714
- Wait a period of time. There should be no warnings about a workspace not being found

Fixes #22714 and Fixes #22736 

Does this update require release notes?
- [ ] Yes
- [x] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidPlot/src/ProjectRecovery.cpp [hunk 1]
```diff
@@ -335,6 +335,8 @@ bool ProjectRecovery::openInEditor(const Poco::Path &inputFolder) {
 void ProjectRecovery::projectSavingThreadWrapper() {
   try {
     projectSavingThread();
+  } catch (Mantid::API::Algorithm::CancelException) {
```
Comment (martyngigg) [inline MantidPlot/src/ProjectRecovery.cpp]: Catch by reference?

File: MantidPlot/src/ProjectRecovery.cpp [hunk 2]
```diff
@@ -435,15 +437,16 @@ void ProjectRecovery::saveWsHistories(const Poco::Path &historyDestFolder) {
   alg->setLogging(false);
 
   for (const auto &ws : wsHandles) {
-    std::string filename = ws;
+    std::string filename = ws->getName();
```
Comment (martyngigg) [inline MantidPlot/src/ProjectRecovery.cpp]: Can we avoid copy the string and use `const auto &`?

Comment (martyngigg) [reply MantidPlot/src/ProjectRecovery.cpp]: Another thought just occurred. What about the case where a workspace name has characters that are invalid in a file path? https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations

Comment (DavidFair) [reply MantidPlot/src/ProjectRecovery.cpp]: It was reported this morning: #22780 
I'll look at addressing it in a future PR

Comment (DavidFair) [reply MantidPlot/src/ProjectRecovery.cpp]: Unfortunately we can't avoid this copy, since we need to append '.py' to the filename

Comment (martyngigg) [reply MantidPlot/src/ProjectRecovery.cpp]: :+1:

**Comment by keeeto** [comment]
Hi @DavidFair , just testing this. When you say a long history, how long is this particular piece of string? I have a history file that is about 2000 lines long, should that suffice? 

Do I just need to make sure that I can click exit between project saving start and end?

**Review by keeeto — APPROVED** [review]
Functional tests all passed.

**Review by martyngigg — APPROVED** [review]
The code changes look sensible. We discussed possible issues with the detach thread living past main() exiting but with the project saving stopping at the beginning of `closeEvent` there should be time for the operation to either complete or get cancelled.

**Review by ianbush — CHANGES_REQUESTED** [review]
This is looking good - I no longer get the segfaults we were discussing earlier.

As we discussed at the ILL home directories are on the network and can be slow. With these changes Mantid hangs on exit for me, while continuing to write the recovery files.

**Comment by DavidFair** [comment]
Talking to @ianbush this issue looks related to writing a large amount of logging to a mounted drive. 

This should be fixed in #22746 , however I've asked him to confirm locally by merging that branch before we proceed with this PR.

**Review by ianbush — APPROVED** [review]
The exiting was actually fine, I was just not testing the branch I thought I was.

**Pull request merged**